src/Guard.php

```diff
@@ -350,125 +335,114 @@ protected function removeFromStorage($name)
-                return $response->withStatus(400)->withHeader('Content-type', 'text/plain')->withBody($body);
-            };
-        }
-        return $this->failureCallable;
-    }
-    
-    /**
-     * Setter for failureCallable
+     * @return ServerRequestInterface
      *
-     * @param mixed $failureCallable Value to set
-     * @return $this
+     * @throws Exception
      */
-    public function setFailureCallable($failureCallable)
+    public function appendNewTokenToRequest(ServerRequestInterface $request): ServerRequestInterface
     {
-        $this->failureCallable = $failureCallable;
-        return $this;
+        $token = $this->generateToken();
+        return $this->appendTokenToRequest($request, $token);
     }
 
     /**
-     * Setter for persistentTokenMode
+     * @param ServerRequestInterface $request
+     * @param $pair
      *
-     * @param bool $persistentTokenMode True to use the same token throughout the session (unless there is a validation error),
-     * false to get a new token with each request.
-     * @return $this
+     * @return ServerRequestInterface
      */
-    public function setPersistentTokenMode($persistentTokenMode)
+    protected function appendTokenToRequest(ServerRequestInterface $request, array $pair): ServerRequestInterface
     {
-        $this->persistentTokenMode = $persistentTokenMode;
-        return $this;
+        $name = $this->prefix . '_name';
+        $value = $this->prefix . '_value';
+        return $request
+            ->withAttribute($name, $pair[$name])
+            ->withAttribute($value, $pair[$value]);
     }
 
     /**
-     * Setter for storageLimit
+     * @param ServerRequestInterface  $request
+     * @param RequestHandlerInterface $handler
      *
-     * @param integer $storageLimit Value to set
-     * @return $this
-     */
-    public function setStorageLimit($storageLimit)
-    {
-        $this->storageLimit = (int)$storageLimit;
-        return $this;
-    }
-
-    /**
-     * Getter for persistentTokenMode
+     * @return ResponseInterface
      *
-     * @return bool
+     * @throws Exception
      */
-    public function getPersistentTokenMode()
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
     {
-        return $this->persistentTokenMode;
-    }
+        if (in_array($request->getMethod(), ['POST', 'PUT', 'DELETE', 'PATCH'])) {
+            $body = $request->getParsedBody();
+            $name = null;
+            $value = null;
 
-    /**
-     * @return string
-     */
-    public function getTokenName()
-    {
-        return isset($this->keyPair[$this->getTokenNameKey()]) ? $this->keyPair[$this->getTokenNameKey()] : null;
+            if (is_array($body)) {
+                $name = $body[$this->prefix . '_name'] ?? null;
+                $value = $body[$this->prefix . '_value'] ?? null;
+            }
+
+            if (!$this->validateToken($name, $value)) {
```

Comment:
In case we change the signature of `validateToken` to allow strings only, then we should write
```php
            if ($name === null || $value === null || !$this->validateToken((string)$name, (string)$value)) {
```